lib/storage/azure.js

```diff
@@ -0,0 +1,260 @@
+ * @param {Object} param0 Azure Service info and container
+ * @param {String} path Remote file path
+ * @param {Function} callback Callback to be called when operation is terminated
+ * @return {any} Result of the callback
+ */
+function removeContainerBlob(blob, path, callback) {
+  blob.svc.deleteBlobIfExists(blob.container, path, function(error, result, response) {
+    if (error) {
+      __log('Cannot delete ' + path + 'on container ' + blob.container);
+    }
+    return callback(error);
+  });
+}
+
+module.exports = function() {
+  var blobSvcs = [];
+
+  var self = {
+    init: function(options, callback) {
+      this.options = options;
+
+      if (!options.replicateClusters || !Array.isArray(options.replicateClusters)) {
+        options.replicateClusters = [];
+      }
+      options.replicateClusters.push({
+        account: options.account,
+        key: options.key,
+        container: options.container
+      });
+      async.each(options.replicateClusters, function(cluster, cb) {
+        createContainer(cluster, options, function(err, svc) {
+          if (err) {
+            return cb(err);
+          }
+          blobSvcs.push({
+            svc: svc,
+            container: cluster.container || options.container
+          });
+          return cb();
+        });
+      }, callback);
+    },
+
+    copyIn: function(localPath, path, options, callback) {
+      gzipme(localPath, false, 'best', function() {
+        path = path[0] === '/' ? path.slice(1) : path;
+        async.each(blobSvcs, function(blobSvc, cb) {
+          createContainerBlob(blobSvc, path, localPath, function(err) {
+            if (err) {
+              __log('! Cannot create blob in container : ' + err);
+            }
+            cb(err);
+          });
+        }, function(err) {
+          if (err) {
+            __log('Cannot create blobs in container: ' + err);
+            return callback(err);
+          }
+          removeLocalBlob(localPath + '.gz', callback);
+        });
+      });
+    },
+
+    copyOut: function(path, localPath, options, callback) {
+      var blob = blobSvcs[0];
+      path = path[0] === '/' ? path.slice(1) : path;
+
+      blob.svc.getBlobToLocalFile(blob.container, path, localPath, function(error, result, response) {
+        if (error) {
+          __log('error azure download file', error);
+          return callback(error);
+        }
+
+        response.localPath = localPath;
+
+        var returnVal = {
+          result: result,
+          response: response
+        };
+
+        return callback(null, returnVal);
+      });
+    },
+
+    remove: function(path, callback) {
+      path = path[0] === '/' ? path.slice(1) : path;
+
+      async.each(blobSvcs, function(blobSvc, cb) {
+        removeContainerBlob(blobSvc, path, cb);
+      }, callback);
+    },
+
+    disable: function(path, callback) {
+      const dPath = utils.getDisabledPath(path, self.options.disabledFileKey);
+      blobSvcs.forEach(blob => {
```

Comment:
This must

Reply:
This must use async.each, like in Ludo's versions. forEach will wind up invoking the callback several times.

Reply:
Use ludo's implementations as a modal here and you should be OK.